File: shimmer.go
package shimmer

import (
	"fmt"
	"strconv"
	"strings"
	"time"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// Default configuration values for the shimmer effect.
const (
	DefaultInterval   = 50 * time.Millisecond // Animation frame rate
	DefaultPeakLight  = 90                    // Maximum lightness percentage (0-100)
	DefaultWaveWidth  = 8                     // Number of characters in the wave
	DefaultWavePause  = 8                     // Pause between wave loops (in characters)
)

// Direction controls which way the shimmer wave moves.
type Direction int

// Shimmer wave direction constants.
const (
	DirectionRight Direction = iota // Wave moves left to right (default)
	DirectionLeft                   // Wave moves right to left
)

// Model creates an animated shimmer effect on text.
// A wave of brightened color sweeps across the text, creating a loading indicator.
type Model struct {
	text       string
	baseColor  string
	isLoading  bool
	position   int
	waveColors []string

	// Configurable options
	interval  time.Duration
	peakLight int
	waveWidth int
	wavePause int
	direction Direction
}

// TickMsg triggers a shimmer animation frame.
type TickMsg time.Time

// Option is a function that configures a Model.
type Option func(*Model)

// WithInterval sets the animation speed (time between frames).
func WithInterval(d time.Duration) Option {
	return func(m *Model) {
		m.interval = d
	}
}

// WithPeakLight sets the maximum lightness (0-100). Higher = brighter shimmer.
func WithPeakLight(percent int) Option {
	return func(m *Model) {
		if percent < 0 {
			percent = 0
		}
		if percent > 100 {
			percent = 100
		}
		m.peakLight = percent
	}
}

// WithWaveWidth sets the width of the shimmer wave in characters.
func WithWaveWidth(width int) Option {
	return func(m *Model) {
		if width < 2 {
			width = 2
		}
		m.waveWidth = width
	}
}

// WithWavePause sets the pause between wave loops (in character positions).
func WithWavePause(pause int) Option {
	return func(m *Model) {
		if pause < 0 {
			pause = 0
		}
		m.wavePause = pause
	}
}

// WithDirection sets the wave direction (left or right).
func WithDirection(dir Direction) Option {
	return func(m *Model) {
		m.direction = dir
	}
}

// New creates a new shimmer model with the given text and base color.
// Color should be a hex color like "#00D787".
func New(text, baseColor string, opts ...Option) Model {
	m := Model{
		text:      text,
		baseColor: baseColor,
		isLoading: true,
		position:  0,
		interval:  DefaultInterval,
		peakLight: DefaultPeakLight,
		waveWidth: DefaultWaveWidth,
		wavePause: DefaultWavePause,
		direction: DirectionRight,
	}

	for _, opt := range opts {
		opt(&m)
	}

	m.waveColors = m.generateWaveColors()
	return m
}

// SetText updates the text being displayed.
func (m Model) SetText(text string) Model {
	m.text = text
	return m
}

// SetLoading enables or disables the shimmer animation.
func (m Model) SetLoading(loading bool) Model {
	m.isLoading = loading
	return m
}

// Init starts the shimmer animation tick.
func (m Model) Init() tea.Cmd {
	if m.isLoading {
		return m.tick()
	}
	return nil
}

// Update handles shimmer animation ticks.
func (m Model) Update(msg tea.Msg) (Model, tea.Cmd) {
	if _, ok := msg.(TickMsg); ok && m.isLoading {
		totalLength := len([]rune(m.text)) + len(m.waveColors) + m.wavePause
		m.position = (m.position + 1) % totalLength
		return m, m.tick()
	}
	return m, nil
}

// View renders the shimmer text with per-character coloring.
func (m Model) View() string {
	baseStyle := lipgloss.NewStyle().Foreground(lipgloss.Color(m.baseColor))

	if !m.isLoading {
		return baseStyle.Render(m.text)
	}

	runes := []rune(m.text)
	var sb strings.Builder

	for i, char := range runes {
		color := m.getCharacterColor(i, len(runes))
		style := lipgloss.NewStyle().Foreground(lipgloss.Color(color))
		sb.WriteString(style.Render(string(char)))
	}

	return sb.String()
}

func (m Model) tick() tea.Cmd {
	return tea.Tick(m.interval, func(t time.Time) tea.Msg {
		return TickMsg(t)
	})
}

func (m Model) getCharacterColor(index, textLen int) string {
	var distance int

	if m.direction == DirectionLeft {
		// Reverse: wave moves right to left
		distance = m.position - (textLen - 1 - index)
	} else {
		// Default: wave moves left to right
		distance = m.position - index
	}

	if distance >= 0 && distance < len(m.waveColors) {
		return m.waveColors[distance]
	}

	return m.baseColor
}

// generateWaveColors creates a high-contrast gradient for the shimmer wave.
// The wave ramps up to peakLight then back down for a smooth pulse.
func (m Model) generateWaveColors() []string {
	r, g, b := parseHexColor(m.baseColor)

	// Build symmetric wave: 0 -> peak -> 0
	steps := m.waveWidth
	if steps < 2 {
		steps = 2
	}

	colors := make([]string, steps)
	mid := steps / 2

	for i := range steps {
		// Calculate distance from edges (0 at edges, 1 at center)
		var ratio float64
		if i <= mid {
			ratio = float64(i) / float64(mid)
		} else {
			ratio = float64(steps-1-i) / float64(steps-1-mid)
		}

		pct := int(ratio * float64(m.peakLight))
		colors[i] = fmt.Sprintf("#%02X%02X%02X",
			lighten(r, pct),
			lighten(g, pct),
			lighten(b, pct),
		)
	}

	return colors
}

// parseHexColor parses a hex color string like "#FFC000" or "FFC000".
func parseHexColor(hex string) (r, g, b int) {
	hex = strings.TrimPrefix(hex, "#")
	if len(hex) != 6 {
		return 0, 215, 135 // Default to green (#00D787) if invalid
	}
	rr, _ := strconv.ParseInt(hex[0:2], 16, 64)
	gg, _ := strconv.ParseInt(hex[2:4], 16, 64)
	bb, _ := strconv.ParseInt(hex[4:6], 16, 64)
	return int(rr), int(gg), int(bb)
}

// lighten blends a color channel toward white (255) by a percentage.
func lighten(value, percent int) int {
	result := value + (255-value)*percent/100
	if result > 255 {
		return 255
	}
	return result
}

File: ts/src/shimmer.ts
import chalk from "chalk";

export const DEFAULT_INTERVAL = 50;
export const DEFAULT_PEAK_LIGHT = 90;
export const DEFAULT_WAVE_WIDTH = 8;
export const DEFAULT_WAVE_PAUSE = 8;

export enum Direction {
  Right = 0,
  Left = 1,
}

export interface Options {
  interval?: number;
  peakLight?: number;
  waveWidth?: number;
  wavePause?: number;
  direction?: Direction;
}

export class Model {
  private text: string;
  private baseColor: string;
  private isLoading: boolean;
  private position: number;
  private waveColors: string[];
  private interval: number;
  private peakLight: number;
  private waveWidth: number;
  private wavePause: number;
  private direction: Direction;
  private timer: ReturnType<typeof setInterval> | null = null;
  private onTick: (() => void) | null = null;

  constructor(text: string, baseColor: string, opts: Options = {}) {
    this.text = text;
    this.baseColor = baseColor;
    this.isLoading = true;
    this.position = 0;
    this.interval = opts.interval ?? DEFAULT_INTERVAL;
    this.peakLight = Math.max(0, Math.min(100, opts.peakLight ?? DEFAULT_PEAK_LIGHT));
    this.waveWidth = Math.max(2, opts.waveWidth ?? DEFAULT_WAVE_WIDTH);
    this.wavePause = Math.max(0, opts.wavePause ?? DEFAULT_WAVE_PAUSE);
    this.direction = opts.direction ?? Direction.Right;
    this.waveColors = this.generateWaveColors();
  }

  setOnTick(callback: () => void): void {
    this.onTick = callback;
  }

  init(): void {
    if (this.isLoading && !this.timer) {
      this.timer = setInterval(() => this.tick(), this.interval);
    }
  }

  stop(): void {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  setText(text: string): this {
    this.text = text;
    return this;
  }

  setLoading(loading: boolean): this {
    this.isLoading = loading;
    if (!loading) {
      this.stop();
    } else if (!this.timer) {
      this.init();
    }
    return this;
  }

  view(): string {
    if (!this.isLoading) {
      return chalk.hex(this.baseColor)(this.text);
    }

    const runes = [...this.text];
    let result = "";

    for (let i = 0; i < runes.length; i++) {
      const color = this.getCharacterColor(i, runes.length);
      result += chalk.hex(color)(runes[i]);
    }

    return result;
  }

  private tick(): void {
    if (!this.isLoading) return;
    const totalLength = [...this.text].length + this.waveColors.length + this.wavePause;
    this.position = (this.position + 1) % totalLength;
    if (this.onTick) {
      this.onTick();
    }
  }

  private getCharacterColor(index: number, textLen: number): string {
    let distance: number;

    if (this.direction === Direction.Left) {
      distance = this.position - (textLen - 1 - index);
    } else {
      distance = this.position - index;
    }

    if (distance >= 0 && distance < this.waveColors.length) {
      return this.waveColors[distance];
    }

    return this.baseColor;
  }

  private generateWaveColors(): string[] {
    const { r, g, b } = parseHexColor(this.baseColor);
    const steps = Math.max(this.waveWidth, 2);
    const colors: string[] = new Array(steps);
    const mid = Math.floor(steps / 2);

    for (let i = 0; i < steps; i++) {
      let ratio: number;
      if (i <= mid) {
        ratio = i / mid;
      } else {
        ratio = (steps - 1 - i) / (steps - 1 - mid);
      }

      const pct = Math.floor(ratio * this.peakLight);
      colors[i] = formatHexColor(lighten(r, pct), lighten(g, pct), lighten(b, pct));
    }

    return colors;
  }
}

function parseHexColor(hex: string): { r: number; g: number; b: number } {
  hex = hex.replace(/^#/, "");
  if (hex.length !== 6) {
    return { r: 0, g: 215, b: 135 };
  }
  return {
    r: parseInt(hex.slice(0, 2), 16),
    g: parseInt(hex.slice(2, 4), 16),
    b: parseInt(hex.slice(4, 6), 16),
  };
}

function formatHexColor(r: number, g: number, b: number): string {
  return `#${r.toString(16).padStart(2, "0").toUpperCase()}${g.toString(16).padStart(2, "0").toUpperCase()}${b.toString(16).padStart(2, "0").toUpperCase()}`;
}

function lighten(value: number, percent: number): number {
  const result = value + ((255 - value) * percent) / 100;
  return Math.min(255, Math.floor(result));
}

export function create(text: string, baseColor: string, opts: Options = {}): Model {
  return new Model(text, baseColor, opts);
}

export { create as New };
